audio-worklet/src/noisegate-script-processor.js

```diff
@@ -0,0 +1,195 @@
+
+      for (let j = 0; j < input.length; j++) {
+        output[j] = weights[j] * input[j];
+      }
+    }
+  }
+
+  /**
+   * Detect level using the difference equation for the Root Mean Squared
+   * value of the input signal.
+   * @param {Float32Array} input audio buffer
+   * @return {Float32Array} the level of the signal
+   */
+  detectLevel_(channel) {
+    // The signal level is determined by filtering high frequency oscillation
+    // with exponential smoothing.
+    // This is equivalent to computing the root-mean-square (RMS) value
+    // of the signal. See http://www.aes.org/e-lib/browse.cfm?elib=16354 for
+    // details.
+    this.envelope_[0] = this.alpha_ * this.lastLevel_ +
+        (1 - this.alpha_) * Math.pow(channel[0], 2);
+
+    for (let j = 1; j < channel.length; j++) {
+      this.envelope_[j] = this.alpha_ * this.envelope_[j - 1] +
+          (1 - this.alpha_) * Math.pow(channel[j], 2);
+    }
+    this.lastLevel_ = this.envelope_[this.envelope_.length - 1];
+    
+    return this.envelope_;
+  }
+
+  /**
+   * Computes an array of weights which determines what samples are silenced.
+   * @param {Float32Array} envelope_ the level of the input
+   * @return {Float32Array} weights numbers in the range 0 to 1 set in
+   *                                accordance with the threshold, the envelope,
+   *                                and attack and release
+   */
+  computeGain_(envelope) {
+    // When attack or release are 0, the weight changes between 0 and 1
+    // in one step.
+    let attackSteps = 1;
+    let releaseSteps = 1;
+    let attackLossPerStep = 1;
+    let releaseGainPerStep = 1;
+
+    // When attack or release are > 0, the associated weight changes between 0
+    // and 1 in the number of steps corresponding to the millisecond attack
+    // or release time parameters.
+    if (this.attack > 0) {
+      attackSteps = Math.ceil(this.context_.sampleRate * this.attack);
+      attackLossPerStep = 1 / attackSteps;
+    }
+    if (this.release > 0) {
+      releaseSteps = Math.ceil(this.context_.sampleRate * this.release);
+      releaseGainPerStep = 1 / releaseSteps;
+    }
+
+    // Compute an array of weights which will be multiplied with the channel.
+    // Based on the detected level and indexes which persist between subsequent
+    // calls to onaudioprocess, the noise gate at iteration i is in one of
+    // four states: open, closed, attacking (between open and closed), or
+    // releasing (between closed and open).
+    for (let i = 0; i < envelope.length; i++) {
+      if (envelope[i] < this.threshold) {
+        const weight = this.lastWeight_ - attackLossPerStep;
+        this.weights_[i] = Math.max(weight, 0);
+      }
+      else {
+        const weight = this.lastWeight_ + releaseGainPerStep;
+        this.weights_[i] = Math.min(weight, 1);
+      }
+      this.lastWeight_ = this.weights_[i];
+    }
+    return this.weights_;
+  }
+
+  getFilterCoefficient_(timeConstant) {
```

Comment:
How about `getAlphaFromTimeConstant`? Also this does not have to be part of the class. Perhaps making it static function would be better?